include/gpiod.h

```diff
@@ -1010,6 +1032,52 @@ const char *
 gpiod_chip_iter_failed_chip(struct gpiod_chip_iter *iter) GPIOD_API;
 
 /**
+ * @brief Iterator structure for GPIO chip bulk objects.
+ */
+struct gpiod_chip_bulk_iter
+{
+	struct gpiod_chip_bulk *bulk;
+	unsigned int index;
+};
+
+/**
+ * @brief Create a new gpiochip bulk iterator.
+ * @return Pointer to a new chip iterator object or NULL if an error occurred.
+ *
+ * This iterator is used to walk over the opened chips of a bulk object without
+ * closing them.
+ */
+struct gpiod_chip_bulk_iter *
+gpiod_chip_bulk_iter_new(struct gpiod_chip_bulk *bulk) GPIOD_API;
+
+/**
+ * @brief Get the next gpiochip handle from the bulk object.
+ * @param iter The gpiochip bulk iterator object.
+ * @return Pointer to an open gpiochip handle or NULL if the next chip can't
+ *         be accessed.
+ *
+ * Internally this routine walks over the previously opened chips of the bulk
+ * object. The returned chips are kept open. If an error occurs or no more
+ * chips are present, the function frees the iterator and returns NULL.
+ */
+struct gpiod_chip *
+gpiod_chip_bulk_iter_next(struct gpiod_chip_bulk_iter *iter) GPIOD_API;
+
+/**
+ * @brief Iterate over all chips in the GPIO chip bulk object
+ * @param bulk GPIO chip bulk object
+ * @param chip Pointer to a GPIO chip object. On each iteration the chip object
+ *             is assigned to this argument without closing the GPIO chip.
+ *
+ * If the chips and its lines are no longer needed, the chips of the GPIO chip
+ * bulk object can be closed by calling gpiod_chip_close_all.
+ */
+#define gpiod_foreach_open_chip(bulk, chip)				\
+	for (struct gpiod_chip_bulk_iter *__iter =			\
+		gpiod_chip_bulk_iter_new((bulk));			\
+	     ((chip) = gpiod_chip_bulk_iter_next(__iter));)
```

Comment:
But the chip variable is not yet assigned at the time of the check, so we'd need to add another call to `gpiod_chip_bulk_iter_next` before and we can't initialize multiple variables at once (iter and chip) unless we statically allocate some `struct { struct gpiod_chip_bulk_iter *iter; struct gpiod_chip *chip } loopstruct = { .iter = gpiod_chip_bulk_iter_new(bulk), .chip = chip}` and use that loopstruct as loop variable, which is not pretty either. Or am I missing something?